module-system/sov-state/src/codec/mod.rs

```diff
@@ -42,3 +43,38 @@ pub trait StateValueCodec<V> {
             .unwrap()
     }
 }
+
+pub trait EncodeLike<Ref: ?Sized, Target> {
+    fn encode_like(&self, borrowed: &Ref) -> Vec<u8>;
+}
+
+// All items can be encoded like themselves
+impl<C, T> EncodeLike<T, T> for C
+where
+    C: StateValueCodec<T>,
+{
+    fn encode_like(&self, borrowed: &T) -> Vec<u8> {
+        self.encode_value(borrowed)
+    }
+}
+
+// In borsh, a slice is encoded the same way as a vector
+impl<T> EncodeLike<[T], Vec<T>> for BorshCodec
```

Comment:
I'm okay with this, but let's document here that vectors and slices don't have *exactly* the same serialization logic starting from Borsh 1.0. Borsh 1.0 disallows zero-sized element types for all collections including `Vec`, but slices are exempt.

https://github.com/near/borsh-rs/blob/09c9295000ced1933e067d094a3cd8fe5507044d/borsh/src/ser/mod.rs#L283

Originally noticed by @bkolad here: https://github.com/Sovereign-Labs/sovereign-sdk/pull/689#discussion_r1301253205

I think we can consider it almost a pathological edge case and still implement the trait.